db/src/diff_set.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+use std::collections::BTreeMap;
+
+/// Witness assertions and retractions, folding (assertion, retraction) pairs into alterations.
+/// Assumes that no assertion or retraction will be witnessed more than once.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
+pub struct DiffSet<K, V> {
+    pub asserted: BTreeMap<K, V>,
+    pub retracted: BTreeMap<K, V>,
+    pub altered: BTreeMap<K, (V, V)>,
+}
+
+impl<K, V> Default for DiffSet<K, V> where K: Ord {
+    fn default() -> DiffSet<K, V> {
+        DiffSet {
+            asserted: BTreeMap::default(),
+            retracted: BTreeMap::default(),
+            altered: BTreeMap::default(),
+        }
+    }
+}
+
+impl<K, V> DiffSet<K, V> where K: Ord {
+    pub fn witness(&mut self, key: K, value: V, added: bool) {
+        if added {
+            if let Some(retracted_value) = self.retracted.remove(&key) {
+                self.altered.insert(key, (retracted_value, value));
+            } else {
+                self.asserted.insert(key, value);
+            }
+        } else {
+            if let Some(asserted_value) = self.asserted.remove(&key) {
+                self.altered.insert(key, (value, asserted_value));
```

Comment:
Oh, I see. If you get the addition before the retraction, retract the old value and assert the new one.